bin/cheritest/cheritest.h

```diff
@@ -151,6 +153,41 @@ void	signal_handler_clear(int sig);
 #define CHERITEST_VERIFY(cond) \
 	CHERITEST_VERIFY2(cond, "%s", "\'" #cond "\' is FALSE!")
 
+#define CHERITEST_CHECK_EQ(type, fmt, a, b, a_str, b_str)	do {	\
+		type __a = (a);						\
+		type __b = (b);						\
+		CHERITEST_VERIFY2(__a == __b, "%s (" fmt ") == %s ("	\
+		    fmt ") failed!", a_str, __a, b_str, __b);		\
+	} while (0)
+
+#define CHERITEST_CHECK_EQ_BOOL(a, b)	\
+	CHERITEST_CHECK_EQ(_Bool, "%d", a, b, __STRING(a), __STRING(b))
+#define CHERITEST_CHECK_EQ_INT(a, b)	\
+	CHERITEST_CHECK_EQ(int, "0x%lx", a, b, __STRING(a), __STRING(b))
+#define CHERITEST_CHECK_EQ_LONG(a, b)	\
+	CHERITEST_CHECK_EQ(long, "0x%lx", a, b, __STRING(a), __STRING(b))
+#define CHERITEST_CHECK_EQ_SIZE(a, b)	\
+	CHERITEST_CHECK_EQ(size_t, "0x%zx", a, b, __STRING(a), __STRING(b))
+
+static inline void
+_cheritest_check_cap_eq(void *__capability a, void *__capability b,
+    const char *a_str, const char *b_str)
+{
+#define CHECK_CAP_ATTR(accessor, fmt)						\
+	CHERITEST_VERIFY2(accessor(a) == accessor(b),				\
+	    __STRING(accessor) "(%s) (" fmt ") == " __STRING(accessor)		\
+	    "(%s) (" fmt ") failed!", a_str, accessor(a), b_str, accessor(b))
+	CHECK_CAP_ATTR(cheri_getaddress, "0x%ld");
+	CHECK_CAP_ATTR(cheri_gettag, "%d");
+	CHECK_CAP_ATTR(cheri_getoffset, "0x%ld");
+	CHECK_CAP_ATTR(cheri_getlength, "0x%ld");
+	CHECK_CAP_ATTR(cheri_getperm, "0x%ld");
+	CHECK_CAP_ATTR(cheri_gettype, "%ld");
+#undef CHECK_CAP_ATTR
+}
+#define CHERITEST_CHECK_EQ_CAP(a, b)	\
+	_cheritest_check_cap_eq(a, b, __STRING(a), __STRING(b));
```

Comment:
Why by field and not appeal to CExEq?  Alternatively, add a note that cap flags are not yet tested?

Reply:
RISC-V currently has no CExEq.

Reply:
Though this indirection through `_cheritest_check_cap_eq ` that duplicates all the `CHERITEST_CHECK_EQ_*` (and has bugs as mentioned above) seems strange. I'd write this as:

```suggestion
	do { \
		CHERITEST_CHECK_EQ_SIZE(cheri_getaddress(a), cheri_getaddress(b)); \
		CHERITEST_CHECK_EQ_BOOL(cheri_gettag(a), cheri_gettag(b)); \
		CHERITEST_CHECK_EQ_SIZE(cheri_getoffset(a), cheri_getoffset(b)); \
		CHERITEST_CHECK_EQ_SIZE(cheri_getlength(a), cheri_getlength(b)); \
		CHERITEST_CHECK_EQ_SIZE(cheri_getperm(a), cheri_getperm(b)); \
		CHERITEST_CHECK_EQ_PTRDIFF(cheri_gettype(a), cheri_gettype(b)); \
	} while (0)
```

provided you define a `CHERITEST_CHECK_EQ_PTRDIFF`. (And you do want size for all those others, not long, since that's how the builtins are defined in Clang, and 32-bit architectures often have `size_t` as an `unsigned int` rather than an `unsigned long` so the format strings would give you warnings on a hypothetical 32-bit CheriBSD-supported CHERI architecture.)

Reply:
The indirection is there to print the argument to the macro instead of a/b (or risk double evaluation). But yeah it's extremely ugly...